registry/consul/registry.go

```diff
@@ -0,0 +1,186 @@
+package consul
+
+import (
+	"context"
+	"fmt"
+	"slices"
+	"sync"
+	"time"
+
+	consul "github.com/hashicorp/consul/api"
+	"github.com/moderntv/cadre/registry"
+	"google.golang.org/grpc/grpclog"
+)
+
+var logger = grpclog.Component("consul_registry")
+
+var _ registry.Registry = &consulRegistry{}
+
+type consulRegistry struct {
+	client *consul.Client
+	// datacenter represents Consul datacenter.
+	datacenter string
+	// refreshPeriod determines period for Consul catalog API call.
+	refreshPeriod time.Duration
+	// aliases is used for mapping gRPC service names to Consul service names.
+	aliases map[string]string
+	mu      sync.RWMutex
+	// services maps service names to currently known service instances.
+	services map[string][]registry.Instance
+}
+
+func NewRegistry(address, datacenter string, aliases map[string]string, refreshPeriod time.Duration) (registry.Registry, error) {
+	config := consul.DefaultConfig()
+	config.Address = address
+	c, err := consul.NewClient(config)
+	if err != nil {
+		return nil, err
+	}
+
+	r := &consulRegistry{
+		client:        c,
+		datacenter:    datacenter,
+		refreshPeriod: refreshPeriod,
+		aliases:       aliases,
+		services:      make(map[string][]registry.Instance),
+	}
+	return r, nil
+}
+
+func (r *consulRegistry) Register(serviceInstance registry.Instance) error {
+	return nil
+}
+
+func (r *consulRegistry) Deregister(serviceInstance registry.Instance) error {
+	return nil
+}
+
+func (r *consulRegistry) Instances(service string) []registry.Instance {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+	is, ok := r.services[service]
+	if !ok {
+		return []registry.Instance{}
+	}
+
+	return is
+}
+
+func (r *consulRegistry) Watch(service string) (<-chan registry.RegistryChange, func()) {
+	// FIX: The change propagation (i.e. the channel communication) has to be somehow reworked.
```

Comment:
Create issue and remove `FIX`.

Reply:
I am certainly making an issue after the request gets merged. I am personally fine with leaving it in the code as a reminder so it hopefully does not get ignored. It will be removed when the issue is solved.

Reply:
Issue: https://github.com/moderntv/cadre/issues/45